packages/server-wallet/src/protocols/defund-channel.ts

```diff
@@ -27,44 +25,37 @@ export class ChannelDefunder {
 
   public async crank(objective: DBDefundChannelObjective, response: WalletResponse): Promise<void> {
     const {targetChannelId: channelId} = objective.data;
-    await this.store.transaction(async tx => {
-      const channel = await this.store.getAndLockChannel(channelId, tx);
-
+    await this.store.lockApp(channelId, async (tx, channel) => {
       if (!channel) {
         this.logger.error(`No channel found for channel id ${channelId}`);
         await this.store.markObjectiveStatus(objective, 'failed', tx);
         return;
       }
 
+      await channel.$fetchGraph('funding', {transaction: tx});
+      await channel.$fetchGraph('chainServiceRequests', {transaction: tx});
+
+      // This if-statement should be removed and test cases should be added.
+      // Defund channel now (in theory) supports Ledger funded channels.
       if (channel.fundingStrategy !== 'Direct') {
         // TODO: https://github.com/statechannels/statechannels/issues/3124
         this.logger.error(`Only direct funding is currently supported.`);
         await this.store.markObjectiveStatus(objective, 'failed', tx);
         return;
       }
 
-      const result = await AdjudicatorStatusModel.getAdjudicatorStatus(tx, channelId);
-
-      const channelHoldings =
-        channel.assetHolderAddress &&
-        (await this.store.getFunding(channelId, makeAddress(channel.assetHolderAddress), tx));
-
-      const outcomePushed = !channelHoldings || BigNumber.from(channelHoldings.amount).isZero();
-      if (result.channelMode === 'Finalized') {
-        if (!outcomePushed) {
-          await ChainServiceRequest.insertOrUpdate(channelId, 'pushOutcome', tx);
-          await this.chainService.pushOutcomeAndWithdraw(result.states[0], channel.myAddress);
-          await this.store.markObjectiveStatus(objective, 'succeeded', tx);
-          response.queueSucceededObjective(objective);
-        } else {
-          this.logger.trace('Outcome already pushed, doing nothing');
-        }
-      } else if (channel.hasConclusionProof) {
-        await ChainServiceRequest.insertOrUpdate(channelId, 'withdraw', tx);
-        await this.chainService.concludeAndWithdraw(channel.support);
-        objective = await this.store.markObjectiveStatus(objective, 'succeeded', tx);
+      const {didSubmitTransaction} = await Defunder.create(
+        this.store,
+        this.chainService,
+        this.logger,
+        this.timingMetrics
+      ).isChannelDefunded(channel, tx);
+
+      // A better methodology is likely to create a Challenge objective that succeeds after a
+      // channel has been defunded (instead of succeeding an objective on transaction submission)
+      if (didSubmitTransaction) {
```

Comment:
Does this mean that after we submit a `challenge` the defund objective has succeeded?

Reply:
It does. I believe that is how the Defund protocol behaves before this refactor.

Reply:
Originally, the defund and challenge submission were two different objectives. When a challenge expired we'd create a defund objective that would be responsible for submitting a `pushOutcomeAndWithdraw` transaction.

If that's still the case I think we could remove `didSubmitTransaction` and change the `Defunder` protocol to just worry about defunding.

Reply:
Gotch! The goal with this PR was to have a pure refactor without changing functionalities of protocols. I agree with you that `didSubmitTransaction` should probably be removed once we implement the new Challenge objective.

At the moment though, unless we want to change the behavior of the Defund objective/protocol, I think we need `didSubmitTransacation`. But maybe I am missing something?